File: Array/905. Sort Array By Parity.py
'''
905. Sort Array By Parity
Array

Description:
Given an array A of non-negative integers, return an array consisting of all the even elements of A, 
followed by all the odd elements of A.
You may return any answer array that satisfies this condition.

Example 1:
Input: [3,1,2,4]
Output: [2,4,3,1]
The outputs [4,2,3,1], [2,4,1,3], and [4,2,1,3] would also be accepted.
'''

# Solution:
class Solution(object):
    def sortArrayByParity(self, A):
        """
        :type A: List[int]
        :rtype: List[int]
        """
        
        '''
        Method: Create two pointers.
                The first pointer sweeps the entire array and searches for the even elements.
                The second pointer counts the number of even numbers. It stops at the odd element positions.
        '''
        count = 0
        for i in range(len(A)):
            if A[i] % 2 == 0:
                A[count], A[i] = A[i], A[count]
                count += 1
        return A

File: Array/905. Sort Array By Parity.cpp
class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& A) {
        
        /* Method 1: create two vectors. A_even contains even elements, A_odd contains odd elements
                     Then insert all odd elements into A_even
        */
        vector<int> A_even, A_odd;
        for (int i=0; i<A.size(); i++) {
            if (A[i]%2==0) {
                A_even.push_back(A[i]);
            }else {
                A_odd.push_back(A[i]);
            }
        }
        for (int j=0; j<A_odd.size(); j++) {
            A_even.push_back(A_odd[j]);
        }
        return A_even;
      
        
        /* Method 2: two pointers.
                     The first pointer sweeps through vector.
                     The second pointer points to the beginning position of odd elements.
        */
        int j=0, tmp;
        for (int i=0; i<A.size(); i++) {
            if (A[i]%2==0) {
                swap(A[j], A[i]);
                j++;
            }
        }
        return A;
    }
};
